src/censure.jl

```diff
@@ -0,0 +1,236 @@
+	in_weight :: Float64
+	out_weight :: Float64
+
+end
+
+type OctagonFilter <: BiFilter
+	
+	m_out :: Integer
+	m_in :: Integer
+	n_out :: Integer
+	n_in :: Integer
+	in_area :: Float64
+	out_area :: Float64
+	in_weight :: Float64
+	out_weight :: Float64
+
+end
+
+type CENSURE <: Detector
+
+	smallest :: Integer
+	largest :: Integer
+	filter_type :: Type
+	filter_stack :: Array
+	responseThreshold :: Number
+	lineThreshold :: Number
+
+end
+
+function OctagonFilter(mo, mi, no, ni)
+	OF = OctagonFilter(mo, mi, no, ni, 0.0, 0.0, 0.0, 0.0)
+	OF.out_area = OF.m_out ^ 2 + 2 * OF.n_out ^ 2 + 4 * OF.m_out * OF.n_out
+	OF.in_area = OF.m_in ^ 2 + 2 * OF.n_in ^ 2 + 4 * OF.m_in * OF.n_in
+	OF.out_weight = 1.0 / (OF.out_area - OF.in_area)
+	OF.in_weight = 1.0 / OF.in_area
+	OF
+end
+
+function BoxFilter(s)
+	BF = BoxFilter(s, 2 * s + 1, 4 * s + 1, (2 * s + 1) ^ 2, (4 * s + 1) ^ 2, 0.0, 0.0)
+	BF.in_weight = 1.0 / BF.in_area
+	BF.out_weight = 1.0 / (BF.out_area - BF.in_area)
+	BF
+end
+
+OctagonFilter_Kernels = [
+						[5, 3, 2, 0],
+						[5, 3, 3, 1],
+						[7, 3, 3, 2],
+						[9, 5, 4, 2],
+						[9, 5, 7, 3],
+						[13, 5, 7, 4],
+						[15, 5, 10, 5]
+						]
+
+BoxFilter_Kernels = [1, 2, 3, 4, 5, 6, 7]
+
+Kernels = Dict(BoxFilter => BoxFilter_Kernels, OctagonFilter => OctagonFilter_Kernels)
+
+function getFilterStack(filter_type::Type, smallest::Integer, largest::Integer)
+	k = Kernels[filter_type]
+	filter_stack = map(f -> filter_type(f...), k[smallest : largest])
+end
+
+getIntegralImage(img, filter_type::BoxFilter) = integral_image(img)
+
+function getIntegralImage(img, filter_type::OctagonFilter)
+	img_shape = size(img)
+	int_img = zeros(img_shape)
+	right_slant_img = zeros(img_shape)
+	left_slant_img = zeros(img_shape)
+
+	int_img[1, :] = cumsum(img[1, :])
+	right_slant_img[1, :] = int_img[1, :]
+	left_slant_img[1, :] = int_img[1, :]
+
+	for i in 2:img_shape[1]
+		sum = 0.0
+		for j in 1:img_shape[2]
+			sum += img[i, j]
+			int_img[i, j] = sum + int_img[i - 1, j]
+			left_slant_img[i, j] = sum
+			right_slant_img[i, j] = sum
+
+			if j > 1 left_slant_img[i, j] += left_slant_img[i - 1, j - 1] end
+			right_slant_img[i, j] += j < img_shape[2] ? right_slant_img[i - 1, j + 1] : right_slant_img[i - 1, j]
+		end
+	end
+	int_img, right_slant_img, left_slant_img
+end
+
+function filterResponse{T}(int_img::AbstractArray{T, 2}, BF::BoxFilter)
+	margin = BF.scale * 2
+	n = BF.scale
+	img_shape = size(int_img)
+	response = zeros(img_shape)
+	R = CartesianRange(CartesianIndex((margin + 1, margin + 1)), CartesianIndex((img_shape[1] - margin, img_shape[2] - margin))) 
+    in_sum = 0.0
+    out_sum = 0.0
+    for I in R
+    	topleft = I + CartesianIndex(- n - 1, - n - 1)
+    	topright = I + CartesianIndex(- n - 1, n)
+    	bottomleft = I + CartesianIndex(n, - n - 1)
+    	bottomright = I + CartesianIndex(n, n)
+    	A = topleft >= CartesianIndex(1, 1) ? int_img[topleft] : 0.0
+    	B = topright >= CartesianIndex(1, 1) ? int_img[topright] : 0.0
```

Comment:
To safe typing you can assign consts before the loop, e.g. `o = one(CartesianIndex{2})` and just use `o`.